secboot/secboot_tpm.go

```diff
@@ -308,6 +321,27 @@ func unlockEncryptedPartitionWithSealedKey(tpm *sb.TPMConnection, name, device,
 	return nil
 }
 
+// AuthKeyFromKernelKeyring obtains the TPM policy update authorization key
+// for the specified disk and device name from the kernel keyring and unlinks
+// the user keyring from the session keyring.
+func AuthKeyFromKernelKeyring(sourceDevice string) ([]byte, error) {
+	keep := false
+	k, err := sbGetActivationDataFromKernel(keyringPrefix, sourceDevice, keep)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get activation data: %v", err)
+	}
+	if err := unlinkUserKeyring(); err != nil {
```

Comment:
Not an expert so bear with me, isn't this modifying the global state of a user (uid 0) keyring? What happens across snapd restarts? Does it affect other services?

Reply:
no, afaiu, what is modified is the session key ring which because of shared is a fresh one created by systemd for the snapd process

Reply:
Yeah, when started by systemd, snapd will have its own session keyring. I did wonder after suggesting this though whether it's safe to assume that snapd always has its own session keyring in all contexts - ie, are there any test scenarios where this code could run with the default (user-session) keyring which can be shared with other processes, and which you wouldn't want to modify?

I wonder whether it might be safer for snapd to just join a new anonymous session keyring, which would have mostly the same effect as unlinking the user keyring (as long as snapd hasn't storing anything in its session keyring at this point) - the user keyring ends up detached from snapd and the previous session keyring that systemd created gets freed by the kernel.

Reply:
The idea is that when snapd restarts it will again be granted access to the user keyring so it can read the contents and then unlink the user keyring from the system session keyring. (I'm not entirely familiar with all the cases and mechanisms used to restart snapd but as long as it's respawned by systemd this should happen.)

Reply:
@chrisccoulson We can change it to join a new anonymous session keyring if you think it's a better idea. We can also do it in a follow-up.